rust/blockstore/src/arrow/block/delta/builder_storage.rs

```diff
@@ -0,0 +1,164 @@
+use std::collections::BTreeMap;
+
+use crate::{arrow::types::ArrowWriteableValue, key::CompositeKey};
+
+pub struct BTreeBuilderStorage<V: ArrowWriteableValue> {
+    storage: BTreeMap<CompositeKey, V>,
+}
+
+impl<V: ArrowWriteableValue + 'static> BTreeBuilderStorage<V> {
+    fn add(&mut self, key: CompositeKey, value: V) {
+        self.storage.insert(key, value);
+    }
+
+    fn delete(&mut self, key: &CompositeKey) -> Option<V> {
+        self.storage.remove(key)
+    }
+
+    fn min_key(&self) -> Option<&CompositeKey> {
+        self.storage.keys().next()
+    }
+
+    fn split_off(&mut self, key: &CompositeKey) -> Self {
+        let split_off = self.storage.split_off(key);
+        Self { storage: split_off }
+    }
+
+    fn len(&self) -> usize {
+        self.storage.len()
+    }
+
+    fn iter<'referred_data>(
+        &'referred_data self,
+    ) -> Box<dyn Iterator<Item = (&'referred_data CompositeKey, &'referred_data V)> + 'referred_data>
+    {
+        Box::new(self.storage.iter())
+    }
+
+    fn into_iter(self) -> Box<dyn Iterator<Item = (CompositeKey, V)>> {
+        Box::new(self.storage.into_iter())
+    }
+}
+
+impl<V: ArrowWriteableValue> Default for BTreeBuilderStorage<V> {
+    fn default() -> Self {
+        Self {
+            storage: BTreeMap::new(),
+        }
+    }
+}
+
+/// This storage assumes that KV pairs are added in order. Deletes are a no-op. Calling `.add()` with the same key more than once is not allowed.
+pub struct VecBuilderStorage<V: ArrowWriteableValue> {
+    storage: Vec<(CompositeKey, V)>,
+}
+
+impl<V: ArrowWriteableValue + 'static> VecBuilderStorage<V> {
+    fn add(&mut self, key: CompositeKey, value: V) {
```

Comment:
can we track the last key and error if not in order?

Reply:
the naive solution seems to result in a significant performance penalty, will investigate to see if I can reduce it

<img width="1059" alt="Screenshot 2024-10-28 at 5 42 43 PM" src="https://github.com/user-attachments/assets/4b20a899-48b7-4a73-9a3a-9fab28278f91">

Reply:
I can't reduce the performance penalty, currently at 20-30%, so going to leave out for now.